### PR TITLE
Remove assertion from jl_sig_throw

### PR DESCRIPTION
**Pull request by vchuravy**
x-ref: https://github.com/JuliaLang/julia/commit/31e76c2dc85#r35077818

### CODE REVIEW

File: src/task.c [hunk 1]
```diff
@@ -438,7 +438,7 @@ void JL_NORETURN throw_internal(jl_value_t *exception JL_MAYBE_UNROOTED)
         jl_longjmp(*ptls->safe_restore, 1);
     JL_GC_PUSH1(&exception);
     jl_gc_unsafe_enter(ptls);
-    if (exception) {
+    if (exception && (ptls->bt_size != 0)) {
```
Comment (c42f) [inline src/task.c]: I guess the exception stack should be able to handle the case of `ptls->bt_size == 0` and still preserve a non-null exception for later analysis.

Comment (c42f) [inline src/task.c]:
```suggestion
    if (exception) {
```

**Review by c42f — COMMENTED** [review]
I need some more context here. In what circumstances can we have `bt_size == 0`? I assume this is valid (or at least unavoidable) in come cases?

I'd argue that we should be preserving the `exception` on the stack, even if we couldn't capture the backtrace.

File: src/task.c (hunk 1 above)
Comment (c42f) [reply src/task.c]: (If it can't handle `bt_size == 0` that's a bug and should be fixed.)

**Comment by vchuravy** [comment]
> I need some more context here. In what circumstances can we have `bt_size == 0`? I assume this is valid (or at least unavoidable) in come cases?

Fair enough the context I have encountered this in is #33255. Looking at it again it originates from: 

```
#0  jl_throw_in_ctx (ptls=0x7f8d01b124a0, e=0x7f8cfd246690 <jl_system_image_data+45685008>, sig=11, sigctx=0x7f8d00b649c0) at /home/vchuravy/src/julia/src/signals-unix.c:176
#1  0x00007f8d08f8c424 in segv_handler (sig=11, info=0x7f8d00b64af0, context=0x7f8d00b649c0) at /home/vchuravy/src/julia/src/signals-unix.c:254
#2  <signal handler called>
#3  0x00007f8d08f8266f in gc_scrub_range (low=0x7f8cee09b000 "", high=0x7f8cee49b000 "") at /home/vchuravy/src/julia/src/gc-debug.c:553
#4  0x00007f8d08f8294c in gc_scrub_task (ta=0x7f8cf608a710) at /home/vchuravy/src/julia/src/gc-debug.c:602
#5  0x00007f8d08f829a3 in gc_scrub () at /home/vchuravy/src/julia/src/gc-debug.c:608
#6  0x00007f8d08f7efd2 in _jl_gc_collect (ptls=0x7f8d01b124a0, full=0) at /home/vchuravy/src/julia/src/gc.c:2812
```

```
$1 = (jl_value_t *) 0x7f8cfd246690 <jl_system_image_data+45685008>
(rr) p jl_(e)
StackOverflowError()
```

https://github.com/JuliaLang/julia/blob/3f5d56abd98cf36dd5fc1eca6e87361f7f9225cf/src/signals-unix.c#L253-L255

Which is a `safe_restore`.

**Review by c42f — APPROVED** [review]
Oh, I see. I did follow your xref originally but I somehow failed to see your comment and I got confused. Sorry about that!

So this is the `safe_restore` codepath... yes, the assert is invalid in that case. I'm fine with just removing it, though it would probably make a tad more sense to change it to

```julia
assert((e && ptls->bt_size != 0) || ptls->safe_restore);
```

if you can be bothered. Either way I'm happy.

**Comment by vchuravy** [comment]
> So this is the safe_restore codepath... yes, the assert is invalid in that case. I'm fine with just removing it, though it would probably make a tad more sense to change it to

I was thinking about that, but it would make more sense to me to add this invariant close to where it is needed.

**Comment by c42f** [comment]
Good point. The tricky thing is that this was trying to check an invariant on data coming out of the signal handlers. Hence having it in sig throw is the first place it could intercept that program state. But I messed it up for the exact reason you describe!

**Pull request merged**